<a href="https://colab.research.google.com/github/Masakinka/python_for_ds_task/blob/main/HW_2_1_%D0%A2%D0%B5%D1%81%D1%82%D1%83%D0%B2%D0%B0%D0%BD%D0%BD%D1%8F_%D1%81%D1%82%D0%B0%D1%82_%D0%B3%D1%96%D0%BF%D0%BE%D1%82%D0%B5%D0%B7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [27]:
import numpy as np
from statsmodels.stats.weightstats import ztest
from scipy.stats import ttest_ind, norm

In [1]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [4]:
np.mean(heights)

170.6969696969697

In [33]:
ztest_Score, p_value = ztest(heights, value = 170, alternative='two-sided')

In [34]:
ztest_Score, p_value

(0.8482235371587619, 0.3963134914218349)

In [35]:
if p_value < 0.05:
    print("Висновок: середній ріст студентів істотно більший за 170 см з рівнем значущості 0.05.")
else:
    print("Висновок: немає статистично значущих доказів, що середній ріст студентів більший за 170 см з рівнем значущості 0.05.")

Висновок: немає статистично значущих доказів, що середній ріст студентів більший за 170 см з рівнем значущості 0.05.


1. Тестова статистика негативна, що вказує на те, що середнє значення у вибірці менше, ніж гіпотетичні 170 см, хоча завдання полягає в тому, щоб перевірити, чи більше середнє зроста. Встановлення альтернативної гіпотези ('two-sided') відповідає негативної тестової статистиці.
2.  Враховуючи завдання та напрямок тесту (односторонній, перевірка на те, чи більше середнє), p-значення близьке до 1.0 (округлене у виведенні 0.8482235371587619) фактично підтримує нульову гіпотезу. Це вказує на те, що середный зрост не значно більший за 170 см.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [10]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [11]:
np.mean(group_a_scores), np.mean(group_b_scores)

(77.08125, 74.73875000000001)

In [12]:
ztest(group_a_scores, group_b_scores, value=0)

(2.357354864751502, 0.01840564855252801)

Статистика для двох вибірок z-критерію становить 2.357354864751502, а відповідне p-value дорівнює 0.01840564855252801.

Оскільки p-value < 0.05, у нас є достатньо доказів, то значення p-value свідчить про те, що середні оцінки між двома групами статистично значимо різняться. Це дозволяє відкинути нульову гіпотезу у тому, що відмінностей немає, і підтвердити, що новий курс ефективніший проти стандартним.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [16]:
# перщі 5 записів з групи А
group_a_scores_small_sample = np.array(group_a_scores[:5])

# Проводимо t-теста для 2 незалежних виборок
# параметра equal_var=False зумовлено потенційно різним розміром та дисперсією вибірок
t_stat, p_value = ttest_ind(group_a_scores_small_sample, group_b_scores, equal_var=False)
p_value

0.3010736990344312

Отримане p-value дорівнює 0.3011, що значно вище за встановлений рівень значущості 0.05. На основі проведеного t-тесту для двох незалежних вибірок, де для групи A було взято лише п'ять записів, та повної вибірки для групи B, не знайдено статистично значущих відмінностей у середніх оцінках між групами. Це вказує на те, що, за попередніми даними, нова освітня програма, випробувана на невеликій вибірці групи A, не показала статистично значущої переваги над стандартною програмою, що використовується в групі B.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [20]:
before_training = np.array([57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77])
after_training = np.array([62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01])

Для аналізу впливу тренінгу на продуктивність працівників за допомогою z-тесту для пов'язаних вибірок, потрібно спочатку визначити різницю між двома наборами даних (до і після тренінгу), а потім перевірити, чи є ця різниця статистично значущою.

In [22]:
# Розрахунок різниці
differences = after_training - before_training
differences

array([ 4.65,  3.03,  5.9 ,  2.47,  8.44,  4.26,  1.29,  0.86,  3.97,
        2.55,  4.42,  3.94,  2.85,  1.3 , -0.03,  2.11,  4.71,  3.42,
        0.51,  3.35,  3.77,  1.23,  3.31,  3.11,  0.72,  3.72,  4.12,
        5.17,  5.11,  0.24])

In [24]:
# Розрахунок середнього і стандартного відхилення різниці
mean_diff = np.mean(differences)
std_diff = np.std(differences, ddof=1)  # ddof=1 для незміщеної оцінки стандартного відхилення
mean_diff, std_diff

(3.1499999999999995, 1.8878924574865183)

In [25]:
# Кількість спостережень
n = len(differences)
n

30

In [29]:
# Z-тест
z_score = mean_diff / (std_diff / np.sqrt(n))
p_value = 2 * (1 - norm.cdf(np.abs(z_score)))  # Two-tailed test
z_score, p_value

(9.13890009623916, 0.0)

Z-score дорівнює 9.1389, що є високим значенням. Це означає, що середнє покращення показників продуктивності віддалене від нуля на 9.1389 стандартних відхилень. Таке значне відхилення вказує на те, що тренінг мав значний вплив на покращення показників.

P-value дорівнює 0.0, що вказує на вкрай малу ймовірність отримати таку або більш значущу різницю між двома вибірками, якщо б припущення про відсутність впливу тренінгу було б вірним. Значення p-value менше порога значущості 0.05 дає нам підстави відкинути нульову гіпотезу.

Висновок:
Тренінг мав статистично значущий позитивний вплив на продуктивність працівників. Результати показують, що після проходження тренінгу показники продуктивності працівників значно покращились. Ці результати можуть бути використані для аргументації користі від тренінгу та його подальшого впровадження серед інших груп співробітників або для подальшого розширення цієї програми тренінгу.